### ISSUE TITLE
Correct way to use Application Insights in v4 and .Net 6

### ISSUE THREAD

**Issue by emilssonn** [Azure/azure-functions-dotnet-worker#760]
From reading different issues, discussions and documentation from the past months, it is still unclear to me how to correctly use/install Application Insights in a .Net isolated function running .Net 6 and v4.

I have tested with the nuget Microsoft.Azure.Functions.Worker.Extensions.ApplicationInsights. The absolute most basic features of application insights works with this nuget. But no dependency tracking etc.

Where can I find the correct documentation for using Application Insights with a .Net isolated function running .Net 6 and v4? Including depedency tracking,  log levels filter, using TelemetryClient and TelemetryConfiguration in code.

**Comment by elsand** [comment]
@emilssonn Microsoft.Azure.Functions.Worker.Extensions.ApplicationInsights seems to be non-working. https://github.com/Azure/azure-functions-applicationinsights-extension/tree/main/src/Worker is literally empty except for a csproj. 

I'm not sure quite sure how to parse https://github.com/Azure/app-service-announcements/issues/346#issuecomment-965840479, but it appears that Microsoft.Azure.Functions.Worker.Extensions.ApplicationInsights isn't supposed to be used at this point.

In any case we're experiencing the same issue as you. Rudimentary trace logging is performed for external calls, but they are not tracked as dependencies (without using Microsoft.Azure.Functions.Worker.Extensions.ApplicationInsights).

**Comment by elsand** [comment]
@emilssonn It appears to work if you follow the steps at https://docs.microsoft.com/en-us/azure/azure-monitor/app/worker-service#using-application-insights-sdk-for-worker-services, that is get the `Microsoft.ApplicationInsights.WorkerService` package and add `services.AddApplicationInsightsTelemetryWorkerService();`  to `ConfigureServices` (`services` being `IServiceCollection`).

**Comment by emilssonn** [comment]
`Microsoft.ApplicationInsights.WorkerService` does indeed seem to work, but not fully. Dependencies are logged, but they are not connected to the requests/other telemetry.

I am 99.9% sure I saw some official documentation about `Microsoft.Azure.Functions.Worker.Extensions.ApplicationInsights` and that it was the official/correct way. But I can no longer find it and the app service announcement might explain that.

But I still cannot find any official documentation about .Net isolated functions running .Net 6 and v4.
https://docs.microsoft.com/en-us/azure/azure-monitor/app/worker-service#using-application-insights-sdk-for-worker-services makes no mention of Azure Functions.

**Comment by fabiocav** [comment]
Leaving this assigned to @brettsam as he is driving the logging enhancements activities.

**Comment by ranouf** [comment]
Hi, so finally what happened with Application Insights for Azure Function V4 and .net 6?

**Comment by SeanFeldman** [comment]
I'm facing a migration but hesitant until know this issue is resolved.
Looking forward to @brettsam's follow-up comments.

**Comment by feverline** [comment]
+1 on this.  I'm facing similar issues with dependency telemetry in dotnet-isolated in functions v4, .net 6.

**Comment by brettsam** [comment]
`Microsoft.Azure.Functions.Worker.Extensions.ApplicationInsights` is deprecated now and we should hide it in nuget. I will follow up on that. 

In short -- we abandoned that approach due to startup performance and I don't believe it would have given you everything you were looking for here anyway.

When you enable Application Insights from the portal (or add the `APPLICATIONINSIGHTS_CONNECTION_STRING` app setting), that turns on logging from the Functions host. It also will enable you to log via the `ILogger` that is passed into your function context in the isolated worker (this sends the logs back to the host to log for you). 

However, there are some other scenarios that this does not cover -- dependency tracking in the worker; access to `TelemetryClient`, etc.

Ideally, the `Microsoft.ApplicationInsights.WorkerService` would "just work" when you used it in the dotnet-isolated worker. We're investigating whether that's possible and will update this issue with our findings. There's a set of logging issues that we're working through right now and this is one of the most important ones.

**Comment by wesselkranenborg** [comment]
+1 on this. We are also facing this issue. We use the `Microsoft.ApplicationInsights.WorkerService` but it's not correlating the dependencies. This is making us hesitant in rolling out isolated worker functions further.

**Comment by emilssonn** [comment]
@brettsam Apart from dependency tracking in the worker and access to `TelemetryClient` scenarios (which almost work when using `Microsoft.ApplicationInsights.WorkerService`, apart from dependencies not correlating to other telemetry). What are the sets of logging issues exactly?
From what I can see the main issue is that `ILogger` categories are not honored, they are changed when logging to Application Insights. They are always changed to `Function.<functionname>.User`. Which makes it impossible to configure logging in any good way. For example if I want some of my code to use Information and some Warning etc. This includes framework code aswell. For example if Information logging is enabled, even `HttpClient` logs are written with category `Function.<functionname>.User.` Is this by design and working as intended? Or are one of the logging issues to make log levels and categories work as they do in other .Net scenarios?

These issues have been discussed in other github issues aswell. The logging issues are a major problem

**Comment by arthurzennig07** [comment]
In my case, Logs were showing in Trace and locally. But the Function's Monitor wasn't logging. 
We recently moved the function from dotnet3.1 to dotnet6 and added dependency injection (including log). 
It started showing logs in Monitor once we added the following configuration in function's startup.cs:
`.ConfigureLogging((ctx,logging) =>
                {
                    if (ctx.HostingEnvironment.IsDevelopment())
                    {
                        logging.AddDebug();
                    }
                    logging.AddConsole();
                    logging.Services.Configure<AzureFileLoggerOptions>(options =>
                    {
                        options.FileName = "az.log";
                        options.FileSizeLimit = 5 * 1024;
                        options.RetainedFileCountLimit = 3;
                    }`

**Comment by vinhent** [comment]
We are having the same issue using `Microsoft.ApplicationInsights.WorkerService` . The dependencies are collected but they are not being correlated. We can see them in App Insights, but they are not showing up underneath the request. While this is being fixed, is there any workaround that could be use to "restore" the  operation context before sending the dependency telemetry, maybe with an `ItelemetryProcessor`?

**Comment by pmonte** [comment]
From what we can see from our tests `_logger.BeginScope("MyScope")` doesn't work in Function, i.e. the log appears in Application Insight but Scope is not set in customDimensions.

**Comment by af4jm** [comment]
we can't get Azure pipelines to build correctly with `Microsoft.ApplicationInsights.WorkerService` ... for .NET 5 projects the work-around is to have the pipeline download the 3.1.x SDK, then 5.0.x ... but with .NET 6 the NuGet restore fails, whether the SDK 3.1 step is there or not, so we're stuck on 5 until this is resolved (we're running isolated process, in case it's relevant)

in both cases the error is `Failed to download package 'Microsoft.NETCore.Targets.3.0.0'` referenced by `WorkerExtensions.csproj`

**Comment by AieatAssam** [comment]
Are there any updates on this issue? Lack of correlated dependencies is a major concern as is lack of documentation on enabling this in general.

**Comment by xtellurian** [comment]
I've just run across this issue as I'm digging into my Azure Function application insights telemetry.

We're running functions v4 (dotnet-isolated) and .net6.0. To be honest, I would expect this to be resolved already, since Az Functions v4 has gone GA

@brettsam can you give us an update on when we'll have access to TelemetryClient etc?

**Comment by rhythmnewt** [comment]
Considering that .net core 3.1 is end of life in December 2022, and .net core 5 is already end of life, this needs to be pushed to the forefront. Requiring organizations to migrate/upgrade to v4/core 6 but losing basic functionality is just terrible.

**Comment by vc4u2c** [comment]
Been having this issue for some time now. We are on Functions v4 and .NET 6. Please resolve this quickly.

**Comment by bbieniek** [comment]
We have used this way: https://github.com/Azure/azure-functions-dotnet-worker/issues/822#issuecomment-1088012705
and it seems to work fine

**Comment by oatsoda** [comment]
This is really starting to become very problematic, to the point we're considering ditching Application Insights and using something else.

We're in a situation where we cannot investigate/analyse live data.

Running .net 6 in Isolated Process functions.

These are the things causing problems:

- Auto dependency tracking not working: https://github.com/Azure/azure-functions-dotnet-worker/issues/619

- Custom dependency tracking (for things like Cosmos Direct TCP mode which would never be tracked automatically) which relies on TelemetryClient
    - Had to register the TelemetryClient manually to work around issue: https://github.com/Azure/azure-functions-dotnet-worker/issues/549
    - These manually logged dependencies _**aren't correlated under the Function requests**_ (not quite sure why, but probably because there's no AppInsights request-scoped context being tracked yet?)

- To get Trace logs to appear correctly we're working around the log level issue in Isolated Process functions by manually wrapping the Logger Factory: https://github.com/Azure/azure-functions-host/issues/4689#issuecomment-533195224 

An update on when all of this is going to be sorted would be appreciated as it's getting to the point where we have to seek alternatives.

**Comment by NeilMacMullen** [comment]
I've just spent 3.5 months going round in circles with customer support on this only to finally discover app-insight dependencies are "not supported" for isolated functions  :-( Could we at least get the documentation improved to make it obvious this is a limitation?

**Comment by SeanFeldman** [comment]
Many have already started the obvious - for a GA-ed product this SDK is in a poor state. With serverless that is hosted and managed by Azure, logging and telemetry is a vital tool for troubleshooting. Without it functioning properly clients are running with blinders on. Detailed documentation, functioning Appinsights integration, and fixed related issues should keep the team up at night. Untill then, it's their customers are the ones that are staying up.

**Comment by marios-siatis** [comment]
Hey guys I have a workaround for structured logging by using serilog, to replace the existing ms logger. 
I think is clean as I only need to modify the startup of the application. 

https://medium.com/@marios.shiatis/addressing-the-structured-logging-application-insights-issues-in-azure-functions-v4-net-6-0-f1f63b99807d

**Comment by Tinathnath** [comment]
It could ! I'll give it a try

**Comment by truptisatardekar** [comment]
+1 for facing the same issue. In my Azure function (dotnet-isolated version 4.0), I used Opentelemetry to track postgresql dependency calls, which show up correctly in AppInsights however they aren't correlated with the parent request.

I tried the below code in Azure Function -> Program.cs -> main -> ConfigureServices

Usings 
using Npgsql;
using OpenTelemetry;
using OpenTelemetry.Resources;
using OpenTelemetry.Trace;

```
services.AddLogging(c =>
            {
           c.AddSerilog(
                new LoggerConfiguration()
                  .ReadFrom.Configuration(context.Configuration)
                  .WriteTo.ApplicationInsights(TelemetryConfiguration.CreateDefault(), TelemetryConverter.Traces, Serilog.Events.LogEventLevel.Information)
                  .WriteTo.File("logs/log.txt", rollingInterval: RollingInterval.Day)
                  .Enrich.WithCorrelationId()
                  .CreateLogger()
              );

              // Enable Tracing for postgresql with OpenTelemetry
              var openTelemetryResourceBuilder = ResourceBuilder.CreateDefault().AddService(serviceName: System.Environment.GetEnvironmentVariable("WEBSITE_SITE_NAME").ToLower());
              var openTelemetryTracerProvider = Sdk.CreateTracerProviderBuilder()
                  .SetResourceBuilder(openTelemetryResourceBuilder)
                  .AddSource(System.Environment.GetEnvironmentVariable("WEBSITE_SITE_NAME").ToLower())
                  .AddNpgsql()
                  .AddAzureMonitorTraceExporter(o =>
                    {
                      o.ConnectionString = context.Configuration.GetValue<string?>("APPLICATIONINSIGHTS_CONNECTION_STRING");
                    })
                  .Build();
              services.AddSingleton(openTelemetryTracerProvider);
            });

            services.AddHttpContextAccessor();
```

**Comment by Tinathnath** [comment]
FYI I started a new project v4 + .net6, when using *in-process* runtime, it works like a charm

**Comment by brettsam** [comment]
The preview packages are now available: https://www.nuget.org/packages/Microsoft.Azure.Functions.Worker.ApplicationInsights.

For usage and explanation, see the PR here (official docs will come when this is out of preview): https://github.com/Azure/azure-functions-dotnet-worker/pull/944.

If you have feedback, bugs, etc, please create a new issue with the details and we'll take a look.

**Comment by LenaVinogradov** [comment]
@brettsam  I have used preview package and sample code provided in the PR link today. I still see all dependencies not linked to the parent. The parentId set in dependencies do not correspond to any telemetry objects logged in our appinsights.

**Comment by brettsam** [comment]
@LenaVinogradov -- can you open a separate issue with exactly what you're seeing and what you expect to see? That'd help us narrow it down. Thanks!

**Comment by LockTar** [comment]
> The preview packages are now available: https://www.nuget.org/packages/Microsoft.Azure.Functions.Worker.ApplicationInsights.

So I'm trying this new preview package locally and in Azure with a .Net 6 function v4 isolated test project. So everything is brand new/green field/new project template. I installed the package and changed the `Program.cs` to:

```C#
var host = new HostBuilder()
    .ConfigureFunctionsWorkerDefaults(builder =>
    {
        builder
            .AddApplicationInsights()
            .AddApplicationInsightsLogger();
    })
    .Build();

host.Run();
```

The azure portal generated application insights for me and configured the function app.
But it's still unclear to me why logging `Debug` and `Trace` messages isn't working. 
What do I need to change? The documentation is not clear.

**Comment by vasylstrogushr** [comment]
> > The preview packages are now available: https://www.nuget.org/packages/Microsoft.Azure.Functions.Worker.ApplicationInsights.
> 
> So I'm trying this new preview package locally and in Azure with a .Net 6 function v4 isolated test project. So everything is brand new/green field/new project template. I installed the package and changed the `Program.cs` to:
> 
> ```cs
> var host = new HostBuilder()
>     .ConfigureFunctionsWorkerDefaults(builder =>
>     {
>         builder
>             .AddApplicationInsights()
>             .AddApplicationInsightsLogger();
>     })
>     .Build();
> 
> host.Run();
> ```
> 
> The azure portal generated application insights for me and configured the function app. But it's still unclear to me why logging `Debug` and `Trace` messages isn't working. What do I need to change? The documentation is not clear.

Hi, 
same behavior for me.
After use this package (Microsoft.Azure.Functions.Worker.ApplicationInsights) I don't have any traces or information logs. I had them when used Microsoft.ApplicationInsights.WorkerService but they were not linked in dependency, now nothing at all.

Azure Function v4, .Net 6 isolated.

**Comment by vasylstrogushr** [comment]
Microsoft.Azure.Functions.Worker.ApplicationInsights looks work now for me.
But still not all dependencies are loaded.
For example(screenshot), call to last service has many internal dependencies, but no one is shown here. But traces and information are from all services. 
![image](https://user-images.githubusercontent.com/19962626/187737773-a374e89c-56ca-4a47-96fa-4ce757d290cc.png)

**Comment by lund76** [comment]
Hello,

I'm also struggling like cracy to get the  trace and debug log work:

# TL;DR:
Using an in process function. The out the box logging works fine (even structures the log a bit). But in isolated I don't get any structure, and Traces and Debugs are not logged. Not in console - not in application insight. With a few config in progam.cs I can get a structured log going - but no way traces and debug.

# The simplest possible solutions:
I've created a [repo](https://github.com/lund76/FuncIsolatedLogginIssue) with what works, and what doesn't

## In Process Example that works
The [host file](https://github.com/lund76/FuncIsolatedLogginIssue/blob/in-process-trace-working/host.json) log level it default set to trace

The [HttpExample.cs](https://github.com/lund76/FuncIsolatedLogginIssue/blob/in-process-trace-working/HttpExample.cs) logs both information, trace and debug:
``` c#
           log.LogInformation("Log Request number {InfoLogCount}", 1);
            log.LogTrace("A little trace in the middle");
            log.LogDebug("A littel debug info in the middle");
            log.LogInformation("Log Request number {InfoLogCount}", 2);
```
And this results in expected logs in both "console"
![image](https://user-images.githubusercontent.com/13216252/190151571-2106e8ed-2797-468e-b85b-ea66bb5b79c5.png)

And application insights:
![image](https://user-images.githubusercontent.com/13216252/190151869-4613dd21-6494-4c56-bb3f-278f6e0b59af.png)

And even structures the log somewhat:
![image](https://user-images.githubusercontent.com/13216252/190152182-5efea26a-1544-4104-9661-4fbf823c6212.png)

## Now using an [isolated approach](https://github.com/lund76/FuncIsolatedLogginIssue/commit/a3f86c28249968e2c0118a244cf4952ad6b3147c)

And with the same log events, **the trace and debug doesn't get logged:**

Not in console:
![image](https://user-images.githubusercontent.com/13216252/190153485-08864ce6-1aba-4fe8-b0e7-9eafaa6a5da9.png)

and not in Application insights:
![image](https://user-images.githubusercontent.com/13216252/190153596-03e0d0cb-21ef-439d-92d8-9395bb526289.png)

The base implementation doesn't handle the structured logging:
![image](https://user-images.githubusercontent.com/13216252/190153860-38640fe0-402c-422c-8b69-9fcdffc03d84.png)

[Adding reference to `Microsoft.Extensions.Logging.ApplicationInsights`](https://github.com/lund76/FuncIsolatedLogginIssue/commit/6462fbfa1487d150e733c138580045f9435307b4)
we get the structure back:
![image](https://user-images.githubusercontent.com/13216252/190154645-469ed12c-48de-4844-ac25-1922bebe5349.png)
[And with the existing providers cleared](https://github.com/lund76/FuncIsolatedLogginIssue/commit/317d1bbdcdf9bedf28697db55e534c35383ab9e5) . The dublicate entries that this would give are gone.

**Comment by olandese** [comment]
At the moment it's a mess. It is totally unclear using a function in Isolated Mode what is the right/correct way to log to Application Insights. 
Which packages and version do we need? 
What configuration needs to be made? 
What code changed we need to apply and where?

**Comment by brettsam** [comment]
Hey everyone -- I'm going to lock this issue down as it's getting hard to track the individual issues. I'll respond to everything that's here now and spin off new issues if needed, but please create a new issue with your questions/bugs from now on.

Thanks for all the feedback!

**Comment by brettsam** [comment]
Reply to: https://github.com/Azure/azure-functions-dotnet-worker/issues/760#issuecomment-1231616471

> The azure portal generated application insights for me and configured the function app. But it's still unclear to me why logging `Debug` and `Trace` messages isn't working. What do I need to change? The documentation is not clear.

@LockTar:

The worker should behave just like any other Console application in .NET. You can adjust logging levels and filters via code, appsettings.json, etc -- https://learn.microsoft.com/en-us/aspnet/core/fundamentals/logging/?view=aspnetcore-6.0#configure-logging. 

Maybe the confusing thing here is that there is now two parts to logging configuration: the host and the worker. The host is controlled by `host.json` and the worker is controlled by the normal .NET mechanisms mentioned above. I called this out here in my initial "docs" (see the last bullet), but we're welcoming feedback: https://github.com/Azure/azure-functions-dotnet-worker/pull/944#issue-1282987627.

Please open a new issue with details if this doesn't help -- or if you don't like it :-) I can go into some reasoning there, if needed.

**Comment by brettsam** [comment]
Reply to: https://github.com/Azure/azure-functions-dotnet-worker/issues/760#issuecomment-1233140398

@vasylstrogushr -- can you share a repro in a new issue? We're not doing anything special when we add the App Insights Logger, so I'd expect it to behave exactly the same way.

One thing that I did realize (and will update shortly) is that the App Insights logger injects an extra filter into the logging system to log only at Warning and above. I spent a while trying to figure out why my Information logs didn't appear anywhere, too: 
- https://learn.microsoft.com/en-us/azure/azure-monitor/app/worker-service#ilogger-logs
- https://github.com/microsoft/ApplicationInsights-dotnet/blob/23181177558f2b2879dc6bf411fdf0b7d3edc479/NETCORE/src/Shared/Extensions/ApplicationInsightsExtensions.cs#L405-L428

**Comment by brettsam** [comment]
Reply to: https://github.com/Azure/azure-functions-dotnet-worker/issues/760#issuecomment-1233198349

@vasylstrogushr -- Can you open a new issue with a small repro to explain the scenario? I can get the right App Insights people to help out there. Are you saying that logs from the separate service are not being correctly correlated?

**Comment by brettsam** [comment]
Reply to: https://github.com/Azure/azure-functions-dotnet-worker/issues/760#issuecomment-1246701048

@lund76 -- you may be hitting the same thing that I mentioned above in https://github.com/Azure/azure-functions-dotnet-worker/issues/760#issuecomment-1277824031 (that host.json only applies to the host). The worker is a separate Console application and it's completely configurable like any other Console application.

Please feel free to file a new issue if this doesn't help or you have feedback. 

I'll give a little overview of our reasoning here: The host.json logging settings are very .NET-specific in their structure. We could load them directly and apply them to the dotnet-isolated application with no problem. But that isn't a great solution for any other language, so we're experimenting with the notion of having the worker control its own logging in its "natural" way and the host being controlled separately via host.json (it is called *host* configuration, after all). This allows you to write filtering code in .NET, apply different log levels to things in host vs worker, etc. And you don't have to think about how two conflicting settings between the host and worker will compose together... the worker ignores host.json completely.

As mentioned above, we welcome feedback here, so feel free to file a new issue. We're trying to rationalize the logging configuration for out-of-proc workers and realizes this can be confusing for those coming from the in-proc world.

**Comment by brettsam** [comment]
https://github.com/Azure/azure-functions-dotnet-worker/issues/760#issuecomment-1264365463

@olandese -- see the "preview doc" at https://github.com/Azure/azure-functions-dotnet-worker/pull/944#issue-1282987627. I *think* it answers all of your questions. If not, please file a new issue with your questions that are unanswered, and I can help out there.
